### ISSUE TITLE
Problems with duplicate functions when running migrate:predeploy 2016-03-12.13

### ISSUE THREAD

**Issue by kskurtveit** [instructure/canvas-lms#804]
### Summary:
In our organization we have 3 running environments with Canvas. One production enviroment, one test and one development. However all of these environments are using rails_env=production so that all environments works the same way. When upgrading our test and dev environments with new releases, we copy the databases we have in production down to our test & dev. 

We started upgrading our test environment from 2016-02-20.12 to 2016-03-12.13 (14), and came across the following errors while running db:migrate:predeploy.

```
TASK: [command su -c "cd /var/www/canvas && RAILS_ENV=production bundle exec rake db:migrate:predeploy" canvas] ***
failed: [t1vlejobs.uib.no] => {"changed": true, "cmd": ["su", "-c", "cd /var/www/canvas && RAILS_ENV=production bundle exec rake db:migrate:predeploy", "canvas"], "delta": "0:00:24.174101", "end": "2016-04-07 17:44:46.742973", "rc": 1, "start": "2016-04-07 17:44:22.568872", "warnings": []}
stderr: DEPRECATION WARNING: Syck is deprecated and support for serialization has been removed. ActiveRecord::Core#yaml_initialize will be removed in 4.1 which will break deserialization support with Syck. (called from <top (required)> at /var/www/canvas/config/application.rb:19)
rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:

PG::DuplicateFunction: ERROR:  function "delayed_jobs_before_insert_row_tr_fn" already exists with same argument types
:       CREATE FUNCTION "delayed_jobs_before_insert_row_tr_fn" () RETURNS trigger AS $$
      BEGIN
        LOCK delayed_jobs IN SHARE ROW EXCLUSIVE MODE;
        IF (SELECT 1 FROM delayed_jobs WHERE strand = NEW.strand LIMIT 1) = 1 THEN
          NEW.next_in_strand := 'f';
        END IF;
        RETURN NEW;
      END;
      $$ LANGUAGE plpgsql;
/var/www/canvas/db/migrate/20110831210257_add_delayed_jobs_next_in_strand.rb:26:in `up'
/var/www/canvas/db/migrate/20120101163452_migrate_queue_to_default_shard.rb:19:in `block in up'
/var/www/canvas/db/migrate/20120101163452_migrate_queue_to_default_shard.rb:17:in `each'
/var/www/canvas/db/migrate/20120101163452_migrate_queue_to_default_shard.rb:17:in `up'
/var/www/canvas/config/initializers/active_record.rb:1261:in `block (2 levels) in migrate'
/var/www/canvas/config/initializers/active_record.rb:1259:in `block in migrate'
/var/www/canvas/config/initializers/active_record.rb:1234:in `each'
/var/www/canvas/config/initializers/active_record.rb:1234:in `migrate'
/var/www/canvas/lib/tasks/canvas.rake:246:in `block (3 levels) in <top (required)>'
ActiveRecord::StatementInvalid: PG::DuplicateFunction: ERROR:  function "delayed_jobs_before_insert_row_tr_fn" already exists with same argument types
:       CREATE FUNCTION "delayed_jobs_before_insert_row_tr_fn" () RETURNS trigger AS $$
      BEGIN
        LOCK delayed_jobs IN SHARE ROW EXCLUSIVE MODE;
        IF (SELECT 1 FROM delayed_jobs WHERE strand = NEW.strand LIMIT 1) = 1 THEN
          NEW.next_in_strand := 'f';
        END IF;
        RETURN NEW;
      END;
      $$ LANGUAGE plpgsql;
/var/www/canvas/db/migrate/20110831210257_add_delayed_jobs_next_in_strand.rb:26:in `up'
/var/www/canvas/db/migrate/20120101163452_migrate_queue_to_default_shard.rb:19:in `block in up'
/var/www/canvas/db/migrate/20120101163452_migrate_queue_to_default_shard.rb:17:in `each'
/var/www/canvas/db/migrate/20120101163452_migrate_queue_to_default_shard.rb:17:in `up'
/var/www/canvas/config/initializers/active_record.rb:1261:in `block (2 levels) in migrate'
/var/www/canvas/config/initializers/active_record.rb:1259:in `block in migrate'
/var/www/canvas/config/initializers/active_record.rb:1234:in `each'
/var/www/canvas/config/initializers/active_record.rb:1234:in `migrate'
/var/www/canvas/lib/tasks/canvas.rake:246:in `block (3 levels) in <top (required)>'
PG::DuplicateFunction: ERROR:  function "delayed_jobs_before_insert_row_tr_fn" already exists with same argument types
/var/www/canvas/db/migrate/20110831210257_add_delayed_jobs_next_in_strand.rb:26:in `up'
/var/www/canvas/db/migrate/20120101163452_migrate_queue_to_default_shard.rb:19:in `block in up'
/var/www/canvas/db/migrate/20120101163452_migrate_queue_to_default_shard.rb:17:in `each'
/var/www/canvas/db/migrate/20120101163452_migrate_queue_to_default_shard.rb:17:in `up'
/var/www/canvas/config/initializers/active_record.rb:1261:in `block (2 levels) in migrate'
/var/www/canvas/config/initializers/active_record.rb:1259:in `block in migrate'
/var/www/canvas/config/initializers/active_record.rb:1234:in `each'
/var/www/canvas/config/initializers/active_record.rb:1234:in `migrate'
/var/www/canvas/lib/tasks/canvas.rake:246:in `block (3 levels) in <top (required)>'
Tasks: TOP => db:migrate:predeploy
(See full trace by running task with --trace)
stdout: 1: production:public
==  MigrateQueueToDefaultShard: migrating =====================================
==  CreateDelayedJobs: migrating ==============================================
-- create_table(:delayed_jobs)
   -> 0.2692s
-- add_index(:delayed_jobs, [:priority, :run_at], {:name=>"delayed_jobs_priority"})
   -> 0.0148s
-- add_index(:delayed_jobs, [:queue], {:name=>"delayed_jobs_queue"})
   -> 0.0142s
==  CreateDelayedJobs: migrated (0.2986s) =====================================

==  AddDelayedJobsTag: migrating ==============================================
-- add_column(:delayed_jobs, :tag, :string)
   -> 0.0013s
-- add_index(:delayed_jobs, [:tag])
   -> 0.0207s
==  AddDelayedJobsTag: migrated (0.0222s) =====================================

==  AddDelayedJobsMaxAttempts: migrating ======================================
-- add_column(:delayed_jobs, :max_attempts, :integer)
   -> 0.0012s
==  AddDelayedJobsMaxAttempts: migrated (0.0014s) =============================

==  AddDelayedJobsStrand: migrating ===========================================
-- add_column(:delayed_jobs, :strand, :string)
   -> 0.0012s
-- add_index(:delayed_jobs, :strand)
   -> 0.0068s
==  AddDelayedJobsStrand: migrated (0.0081s) ==================================

==  CleanupDelayedJobsIndexes: migrating ======================================
-- remove_index(:delayed_jobs, {:name=>"delayed_jobs_queue"})
   -> 0.0257s
-- remove_index(:delayed_jobs, {:name=>"delayed_jobs_priority"})
   -> 0.0033s
==  CleanupDelayedJobsIndexes: migrated (0.0319s) =============================

==  OptimizeDelayedJobs: migrating ============================================
-- create_table(:failed_jobs)
   -> 0.0533s
-- remove_index(:delayed_jobs, {:name=>"get_delayed_jobs_index"})
   -> 0.0038s
-- remove_index(:delayed_jobs, [:strand])
   -> 0.0024s
-- add_index(:delayed_jobs, ["run_at", "queue", "locked_at", "strand", "priority"], {:name=>"index_delayed_jobs_for_get_next"})
   -> 0.0228s
-- add_index(:delayed_jobs, ["strand", "id"], {:name=>"index_delayed_jobs_on_strand"})
   -> 0.0198s
==  OptimizeDelayedJobs: migrated (0.1139s) ===================================

==  AddDelayedJobsNextInStrand: migrating =====================================
-- remove_index(:delayed_jobs, {:name=>"index_delayed_jobs_for_get_next"})
   -> 0.0030s
-- add_column(:delayed_jobs, :next_in_strand, :boolean, {:default=>true, :null=>false})
   -> 0.1648s
-- execute("      CREATE FUNCTION \"delayed_jobs_before_insert_row_tr_fn\" () RETURNS trigger AS $$\n      BEGIN\n        LOCK delayed_jobs IN SHARE ROW EXCLUSIVE MODE;\n        IF (SELECT 1 FROM delayed_jobs WHERE strand = NEW.strand LIMIT 1) = 1 THEN\n          NEW.next_in_strand := 'f';\n        END IF;\n        RETURN NEW;\n      END;\n      $$ LANGUAGE plpgsql;\n")

FATAL: all hosts have already failed -- aborting
```

To make sure we always use the same code in our environments we upgrade Canvas by using our own RPMs. This procedure has worked well for nearly 6 months in production and now it is the first time I've seen this happening. Our dev enviroment didn't suffer from this problem, and in all environments we are using separate canvas and canvas_queue databases.

### Steps to reproduce:
Unsure why it happened in the first place. Cannot reproduce in vagrant-environment. What I have done to fix the problem in our test enviroment was to delete the functions / triggers it complained about and let the predeploy recreate them as needed. This went well, but I'm unsure if this is a change in the way things are structured? Would appreciate some input on this..

Is it safe to delete those functions / triggers in the first place? If yes, should I be prepared to do this again in the near future?

### Expected behavior:
Should have gone through with the migrate:predeploy without having to delete the functions / triggers in the database. Never had a problem with this before, running nearly for 6 months in production. Any thoughts?

**Comment by ccutrer** [comment]
Separate database for the queue is no longer supported by Canvas, and this migration is trying to seamlessly copy your data over to the main database. I'm guessing sometime in the past you accidentally ran migrations or something (in prod) without the queue section, and the delayed_jobs tables got created on the main database. At which point you dropped them (I know they're not there now, because this migration specifically does nothing if the table already exists), but didn't drop the corresponding trigger functions. Just dropping the trigger functions in the main database prior to running this migration is an acceptable solution for your current state.

**Comment by kskurtveit** [comment]
Hi,

Just to let you know when this migration ran in prod just recently when we upgraded that environment it went through without any errors at all. There has been no transaction towards the separate queue databases in any of our environments so this ticket can be closed for now. Thanks for your assistance on this issue.

**Comment by n3rve** [comment]
Hi @kskurtveit: How did you resolve this? I'm upgrading a fairly old version of Canvas to the latest release and I'm encountering the same error:
`PG::DuplicateFunction: ERROR:  function "delayed_jobs_before_insert_row_tr_fn" already exists with same argument types`

**Comment by kskurtveit** [comment]
> Just dropping the trigger functions in the main database prior to running this migration is an acceptable solution for your current state.

So before running the upgrade do: 
```
DROP FUNCTION delayed_jobs_after_delete_row_tr_fn () CASCADE
DROP FUNCTION delayed_jobs_before_insert_row_tr_fn () CASCADE
DROP FUNCTION half_md5_as_bigint ( character varying) CASCADE
```
These triggers should only exist in the canvas database and not in the queue database. Run the drop functions and it should work.

**Comment by n3rve** [comment]
Hi @kskurtveit 

Thank you very much for responding. I've been struggling with this for days.

These are the steps I've taken:
1. I dropped the database and restored a fresh dump from the old canvas installation. 
2. Executed the DROP FUNCTION statements against the database
3. Ran migrations `RAILS_ENV=development bundle exec rake db:migrate` and it failed with this error: P**G::UndefinedFunction: ERROR:  function half_md5_as_bigint(character varying) does not exist**
4. Thinking this must be a result of the last DROP FUNCTION statement, I repeated steps one to 3 retaining the ***half_md5_as_bigint*** function and it failed with a new error ***PG::DuplicateTable: ERROR:  relation "delayed_jobs" already exists***
5. I logged in via PSQL and dropped the `delayed_jobs` and `failed_jobs` tables, then ran the migrations again.
6. This time, all migrations completed successfully. I also ran the rake tasks to compile assets and load notifications, both completed successfully. 

Thank you for your help, I made some progress. However, I now have a 500 error when visiting the installation. If I switch ***database.yml*** to use the former database, all is fine but using the just upgraded DB produces a 500 error and Canvas' production.log file is so disorganized, I can barely make sense of it. 

This is my **production.log** file starting up after the upgrade: http://termbin.com/e1fe
